rust/connlib/tunnel/src/tests/buffered_transmits.rs

```diff
@@ -0,0 +1,139 @@
+use super::sim_net::Host;
+use snownet::Transmit;
+use std::{
+    cmp::Reverse,
+    collections::BinaryHeap,
+    time::{Duration, Instant},
+};
+
+/// A buffer for network packets that need to be handled at a certain point in time.
+#[derive(Debug, Clone, Default)]
+pub(crate) struct BufferedTransmits {
+    // Transmits are stored in reverse ordering to emit the earliest first.
+    inner: BinaryHeap<Reverse<ByTime<Transmit<'static>>>>,
+}
+
+#[derive(Clone, Debug, PartialEq, PartialOrd, Eq, Ord)]
+struct ByTime<T> {
+    at: Instant,
+    value: T,
+}
+
+impl BufferedTransmits {
+    /// Pushes a new [`Transmit`] from a given [`Host`].
+    pub(crate) fn push_from<T>(
+        &mut self,
+        transmit: impl Into<Option<Transmit<'static>>>,
+        sending_host: &Host<T>,
+        now: Instant,
+    ) {
+        let Some(transmit) = transmit.into() else {
+            return;
+        };
+
+        if transmit.src.is_some() {
+            self.push(transmit, sending_host.latency(), now);
+            return;
+        }
+
+        // The `src` of a [`Transmit`] is empty if we want to send if via the default interface.
+        // In production, the kernel does this for us.
+        // In this test, we need to always set a `src` so that the remote peer knows where the packet is coming from.
+
+        let Some(src) = sending_host.sending_socket_for(transmit.dst.ip()) else {
+            tracing::debug!(dst = %transmit.dst, "No socket");
+
+            return;
+        };
+
+        self.push(
+            Transmit {
+                src: Some(src),
+                ..transmit
+            },
+            sending_host.latency(),
+            now,
+        );
+    }
+
+    pub(crate) fn push(
+        &mut self,
+        transmit: impl Into<Option<Transmit<'static>>>,
+        latency: Duration,
+        now: Instant,
+    ) {
+        let Some(transmit) = transmit.into() else {
+            return;
+        };
+
+        debug_assert!(transmit.src.is_some(), "src must be set for `push`");
```

Comment:
```suggestion
        assert!(transmit.src.is_some(), "src must be set for `push`");
```

for a test I think it doesn't make sense to use a `debug_assert`?

Reply:
I don't really mind. I've been treating these parts of the test like a library and thus use `debug_assert` for invariants that should be maintained.